src/main/scala/org/zalando/spearheads/innkeeper/api/JsonDtos.scala

```diff
@@ -2,59 +2,84 @@ package org.zalando.spearheads.innkeeper.api
 
 import java.time.LocalDateTime
 
-import org.zalando.spearheads.innkeeper.api.Endpoint.{ EndpointType, Https, Protocol, ReverseProxy }
-import org.zalando.spearheads.innkeeper.api.PathMatcher.MatcherType
-
 import scala.collection.immutable.Seq
 
 /**
  * @author dpersa
  */
-case class Route(id: Long, route: NewRoute, createdAt: LocalDateTime,
-                 deletedAt: Option[LocalDateTime] = None)
+sealed trait RouteName {
+  def name: String
+}
 
-case class Endpoint(hostname: String, path: Option[String] = None,
-                    port: Option[Int] = Some(443),
-                    protocol: Option[Protocol] = Some(Https),
-                    endpointType: Option[EndpointType] = Some(ReverseProxy))
+case class ValidRouteName private[api] (name: String) extends RouteName
 
-object Endpoint {
+case class InvalidRouteName private[api] (name: String) extends RouteName
+
+object RouteName {
+  def apply(name: String) = "[A-Z][A-Z0-9_]*".r.unapplySeq(name) match {
+    case Some(n) => ValidRouteName(name)
+    case None    => InvalidRouteName(name)
+  }
+}
 
-  sealed trait EndpointType
+sealed trait Route {
+  def name: RouteName
 
-  case object ReverseProxy extends EndpointType
-  case object PermanentRedirect extends EndpointType
+  def route: NewRoute
 
-  sealed trait Protocol
+  def activateAt: Option[LocalDateTime]
 
-  case object Http extends Protocol
-  case object Https extends Protocol
+  def description: Option[String]
 }
 
-case class Error(status: Int,
-                 title: String,
-                 errorType: String,
-                 detail: Option[String] = None)
+case class RouteIn(name: RouteName,
+                   route: NewRoute,
+                   activateAt: Option[LocalDateTime],
+                   description: Option[String] = None) extends Route
+
+case class RouteOut(id: Long,
+                    name: RouteName,
+                    route: NewRoute,
+                    createdAt: LocalDateTime,
+                    activateAt: Option[LocalDateTime],
+                    description: Option[String] = None,
+                    deletedAt: Option[LocalDateTime] = None) extends Route
+
+case class NewRoute(matcher: Matcher,
+                    filters: Option[Seq[Filter]] = Some(Seq.empty),
+                    endpoint: Option[String] = None)
+
+case class Matcher(hostMatcher: Option[String] = None,
+                   pathMatcher: Option[PathMatcher] = None,
+                   methodMatcher: Option[String] = None,
+                   headerMatchers: Option[Seq[HeaderMatcher]] = Some(Seq.empty))
```

Comment:
if you don't have it as an option, you are obliged to put it into the json:
``` json
"matcher": {
  "headerMatcher": []
}
```
as opposed to:
``` json
"matcher": {
}
```